cmd/secrets/providers/providers.go

```diff
@@ -0,0 +1,68 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// +build secrets
+
+package providers
+
+import (
+	"fmt"
+	"time"
+
+	s "github.com/DataDog/datadog-agent/pkg/secrets"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+)
+
+type SecretsProvider interface {
+	// Get Fetches a secret
+	// Note that secretID has a different format on each secret provider. In the
+	// file provider, it's a path, whereas in the kubernetes secret one it has
+	// the following format: "namespace/name/key".
+	Get(secretID string) s.Secret
+}
+
+type Prefix string
+
+const (
+	file      Prefix = "file"
+	k8sSecret Prefix = "k8s_secret"
+)
+
+func (prefix Prefix) IsValid() bool {
+	return prefix == file || prefix == k8sSecret
+}
+
+type Selector struct {
+	File      *FileProvider
+	K8sSecret *K8sSecretProvider
```

Comment:
Do we need those to be struct ? both of them could be a simple function. We could simplify the interface and code by simply having a map of prefix to function.

Since this is meant to be used as a command we won't save any state between call.

WDYT?

Reply:
I've tried a different approach with some of the ideas that you mentioned in this comment and the one below.
The  `providers` is now much simpler to use for the caller: `GetSecret(prefix Prefix, secretId string)`.

I think there are 2 important constraints when implementing this:
- We shouldn't try to initialize a Kubernetes client until we find a "k8s_secret" prefix. This is for users that are only interested in other providers ("file" for now, but could be others in the future).
- We need an easy way to inject the Selector for testing. That's why I kept the `Selector` struct.

Reply:
Overall I think the `GetSecret` call is simpler but we complexified everything else.

> We shouldn't try to initialize a Kubernetes client until we find a "k8s_secret" prefix. This is for users that are only interested in other providers ("file" for now, but could be others in the future).

A basic function would not be called before we detect a k8s_secret so we would only initialize the client when needed.

> We need an easy way to inject the Selector for testing. That's why I kept the Selector struct.

A map of function handler would allow injecting test Selector.

---

Overall I don't have a very strong opinion on this, but I feel that the code could be simpler/shorter and therefore easier to maintain.